fmralign/template_alignment.py

```diff
@@ -0,0 +1,393 @@
+
+
+def euclidian_mean(imgs, masker, scale_template=False):
+    """ Make the euclidian average of images.
+
+    Parameters
+    ----------
+    imgs: list of Niimgs
+        Each img is 3D by default, but can also be 4D.
+    masker:
+
+    scale_template: boolean
+        If true, the returned average is scaled to have the average norm of imgs
+        If false, it will usually have a smaller norm than initial average
+        because noise will cancel across images
+
+    Returns
+    -------
+    average_img: Niimg
+        Average of imgs, with same shape as one img
+    """
+    masked_imgs = [masker.transform(img) for img in imgs]
+    average_img = np.mean(masked_imgs, axis=0)
+    scale = 1
+    if scale_template:
+        X_norm = 0
+        for img in masked_imgs:
+            X_norm += np.linalg.norm(img)
+        X_norm /= len(masked_imgs)
+        scale = X_norm / np.linalg.norm(average_img)
+    average_img *= scale
+
+    return masker.inverse_transform(average_img)
+
+
+def _align_images_to_template(imgs, template, alignment_method,
+                              n_pieces, clustering_method, n_bags, masker,
+                              memory, memory_level, n_jobs, verbose):
+    '''Convenience function : for a list of images, return the list
+    of estimators (PairwiseAlignment instances) aligning each of them to a
+    common target, the template. All arguments are used in PairwiseAlignment
+    '''
+    aligned_imgs = []
+    for img in imgs:
+        piecewise_estimator = \
+            PairwiseAlignment(n_pieces=n_pieces,
+                              alignment_method=alignment_method,
+                              clustering_method=clustering_method, n_bags=n_bags,
+                              mask=masker, memory=memory,
+                              memory_level=memory_level,
+                              n_jobs=n_jobs, verbose=verbose)
+        piecewise_estimator.fit(img, template)
+        aligned_imgs.append(piecewise_estimator.transform(img))
+    return aligned_imgs
+
+
+def create_template(imgs, n_iter, scale_template, alignment_method, n_pieces,
+                    clustering_method, n_bags, masker, memory, memory_level,
+                    n_jobs, verbose):
+    '''Create template through alternate minimization.  Compute iteratively :
+        * T minimizing sum(||R_iX_i-T||) which is the mean of aligned images (RX_i)
+        * align initial images to new template T
+            (find transform R minimizing ||R_iX_i-T|| for each img X_i)
+
+
+        Parameters
+        ----------
+        alignment_method: string
+            Algorithm used to perform alignment between X_i and Y_i :
+            * either 'identity', 'scaled_orthogonal', 'ridge_cv',
+                'permutation', 'diagonal'
+            * or an instance of one of alignment classes
+                (imported from functional_alignment.alignment_methods)
+        n_pieces: int, optional (default = 1)
```

Comment:
For this an other arguments, the defualt parameters are not specified in the function signature.